### PR TITLE
worker: add unit test cases

### PR DESCRIPTION
**Pull request by liuzix**
- Add unit test cases for `BaseWorker`.
- Abstracted out `pkg/clock` for mocking the clock.
- Fixed bugs in `BaseWorker`.

### CODE REVIEW

File: lib/worker.go [hunk 1]
```diff
@@ -312,28 +368,31 @@ func (m *masterClient) HandleHeartbeat(msg *HeartbeatPongMessage) {
...
+	if msg.Epoch < m.masterEpoch {
+		log.L().Info("epoch does not match, ignore stale heartbeat",
 			zap.Any("msg", msg),
 			zap.Int64("master-epoch", m.masterEpoch))
 		return
 	}
+
+	if msg.Epoch > m.masterEpoch {
+		m.masterEpoch = msg.Epoch
+	}
```
Comment (amyangfei) [inline lib/worker.go:380]: Should `msg.Epoch == m.masterEpoch` be never happened.

File: lib/worker_test.go [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+
+	worker.On("InitImpl", mock.Anything).Return(nil)
+	worker.On("Status").Return(WorkerStatus{
+		Code: WorkerStatusNormal,
+	}, nil)
+	err := worker.Init(ctx)
+	require.NoError(t, err)
+
+	worker.clock.(*clock.Mock).Add(defaultTimeoutConfig.workerHeartbeatInterval + 1*time.Second)
+	worker.clock.(*clock.Mock).Add(defaultTimeoutConfig.workerHeartbeatInterval + 1*time.Second)
```
Comment (amyangfei) [inline lib/worker_test.go:48]: These two lines are the same, is it by design?

File: pkg/clock/clock.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+type Clock interface {
+	After(d time.Duration) <-chan time.Time
+	AfterFunc(d time.Duration, f func()) *Timer
+	Now() time.Time
+	Since(t time.Time) time.Duration
+	Sleep(d time.Duration)
+	Tick(d time.Duration) <-chan time.Time
+	Ticker(d time.Duration) *Ticker
+	Timer(d time.Duration) *Timer
```
Comment (amyangfei) [inline pkg/clock/clock.go:26]: What about reusing the interface `clock.Clock` from github.com/benbjohnson/clock for these APIs.

File: lib/worker.go [hunk 4]
```diff
@@ -147,7 +172,13 @@ func (w *BaseWorker) MetaKVClient() metadata.MetaKV {
 	return w.metaKVClient
 }
 
-func (w *BaseWorker) startBackgroundTasks(ctx context.Context) {
+func (w *BaseWorker) startBackgroundTasks() {
+	ctx, cancel := context.WithCancel(context.Background())
```
[the remark this answers is not in the document]
Comment (liuzix) [reply lib/worker.go:176]: Passing a context here will cause additional implicit assumptions about what context the `Executor` is passing us. We are expecting a long-lasting context, so I think using `context.Background()` is safer.

File: lib/worker.go (hunk 1 above)
Comment (liuzix) [reply lib/worker.go:380]: This should be the norm.

File: lib/worker_test.go (hunk 2 above)
Comment (liuzix) [reply lib/worker_test.go:48]: Yes.

Comment (liuzix) [reply lib/worker_test.go:48]: It seems that the mock clock's ticker needs this to trigger a tick.

**Review by amyangfei — APPROVED** [review]

**Pull request merged**